rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala

```diff
@@ -85,6 +107,132 @@ class ReportingServiceImpl(
     }
   }
 
+}
+
+trait CachedFindRuleNodeStatusReports extends ReportingService with CachedRepository with Loggable {
+
+  /**
+   * underlying service that will provide the computation logic
+   */
+  def defaultFindRuleNodeStatusReports: DefaultFindRuleNodeStatusReports
+
+  def nodeInfoService: NodeInfoService
+
+  /**
+   * The cache is managed node by node.
+   * A missing nodeId mean that the cache wasn't initialized for
+   * that node.
+   */
+  private[this] var cache = Map.empty[NodeId, Set[RuleNodeStatusReport]]
+
+  private[this] def cacheToLog(c: Map[NodeId, Set[RuleNodeStatusReport]]): String = {
+    //display compliance value and expiration date.
+    c.map { case (nodeId, reports) =>
+
+      val reportsString = reports.map { r =>
+        val run = r.agentRunTime match {
+          case None => "no agent run"
+          case Some(r) => s"last run@${r}"
+        }
+        val cfid = r.configId match {
+          case None => "no config id"
+          case Some(c) => s"configId:${c.value}"
+        }
+        s"${r.ruleId.value}/${r.serial}[exp:${r.expirationDate}][${run}][${cfid}]${r.compliance.toString}"
+      }.mkString("\n  ", "\n  ", "")
+
+      s"node: ${nodeId.value}${reportsString}"
+    }.mkString("\n", "\n", "")
+  }
+
+  /**
+   * For the nodeIds in parameter, check that the cache is:
+   * - initialized, else go find missing rule node status reports (one time for all)
+   * - none reports is expired, else switch its status to "missing" for all components
+   */
+  def checkAndUpdateCache(nodeIds: Set[NodeId]) : Box[Map[NodeId, Set[RuleNodeStatusReport]]] = this.synchronized {
+    if(nodeIds.isEmpty) {
+      Full(Map())
+    } else {
+      val now = DateTime.now
+
+      /*
+       * Three cases: 1/ cache does not exist, 2/ cache exists but expiration date expired, 3/ cache does note exists
+       * For simplicity (and keeping computation logic elsewhere than in a cache that already has its cache logic
+       * to manage), we will group 2 and 3, but it is well noted that it seems that a RuleNodeStatusReports that
+       * expired could be computed without any more logic than "every thing is missing".
+       *
+       */
+
+      val (expired, upToDate) = nodeIds.partition { id =>  //two group: expired id, and up to date info
```

Comment:
expired may not be the best name, has the comment explain nicely that it is not only expired
missingOrExpired ?